File: strings/longest-valid-parentheses/longest-valid-parentheses.ts
/**
 * @param {string} s
 * @return {number}
 */
function longestValidParentheses(s: string): number {
  let max: number = 0;
  let opening: number = 0;
  let closing: number = 0;

  for (let i = 0; i < s.length; i++) {
    const curr: string = s[i];

    if (curr === '(') {
      opening++;
    } else {
      closing++;
      if (closing > opening) {
        closing = 0;
        opening = 0;
      } else if (closing === opening) {
        max = Math.max(max, closing + opening);
      }
    }
  }

  if (closing < opening) {
    opening = 0;
    closing = 0;

    for (let i = s.length - 1; i >= 0; i--) {
      const curr: string = s[i];

      if (curr === '(') {
        opening++;

        if (opening > closing) {
          closing = 0;
          opening = 0;
        } else if (closing === opening) {
          max = Math.max(max, opening + closing);
        }
      } else {
        closing++;
      }
    }
  }
  return max;
}

File: strings/longest-valid-parentheses/longest-valid-parentheses.go
package longestvalidparentheses

func longestValidParentheses(s string) int {
	max := 0
	opening := 0
	closing := 0

	for i := 0; i < len(s); i++ {
		curr := s[i]

		if curr == '(' {
			opening++
		} else {
			closing++

			if closing > opening {
				closing, opening = 0, 0
			} else if closing == opening {
				sum := opening + closing
				if max < sum {
					max = sum
				}
			}
		}
	}

	if opening > closing {
		opening, closing = 0, 0

		for i := len(s) - 1; i >= 0; i-- {
			curr := s[i]

			if curr == '(' {
				opening++
				if closing < opening {
					closing, opening = 0, 0

				} else if closing == opening {
					sum := closing + opening
					if sum > max {
						max = sum
					}
				}
			} else {
				closing++
			}
		}
	}
	return max
}
